### DISCUSSION TITLE
Make part of PDF start on odd or even page

### DISCUSSION THREAD

**Discussion by alex-webmaster** [dompdf/dompdf#2333] [General]
I convert html to pdf.
Is it possible to make a part of the PDF start only on an odd or even page, for example, using the page-break-after parameter?
Thanks!

**Comment by bsweeney** [comment]
Ideally Dompdf would support left/right for the page-break properties. In which case to start a section of content on an even numbered page you should be able to do something like:

```
<div style="page-break-before: even;">
start on even page
</div>
```

But while left/right are currently supported they are treated the same as "always" (allowed per the spec). I created an issue to enhance left/right support to better accommodate the definition of those values. Follow #2337.

You may be able to accommodate this using callbacks, but I'd have to look into it more.

**Comment by alex-webmaster** [comment]
Can i use code like this?
```
<script type="text/php">
    if ( isset($pdf) ) {
        if($PAGE_NUM%2 != 0)
        {
            echo '<div style="page-break-after: always;"></div>';
        }
    }
</script>
```
but 'echo' not work, maybe there is some method in $pdf object?

I tried the `$pdf->new_page()` method, but when using it, the cursor on the next page jumps to the line number where the text ended on the previous page. I.e. if I add text on the next page, it doesn't come first line when using $pdf->new_page(). Maybe there is something like `$pdf->echo('<div style="page-break-after: always;"></div>')` ?

**Comment by bsweeney** [comment]
By the time Dompdf is processing any embedded script the frame tree has already been constructed. Meaning you can't echo HTML into the document like that. The embedded script is primarily for interacting with the PDF renderer (CPDF, PDFLib) rather than Dompdf.

As you saw you can force a new page but it doesn't reset the current positioning since it's interacting with the PDF backend rather than Dompdf.

This is where callbacks would come into play. I don't work enough with them to recall off the top of my head what this would require so I'd have to play around (and maybe put together some documentation at the same time). You can see a [callback in action in the unit tests](https://github.com/dompdf/dompdf/blob/v1.0.0/tests/DompdfTest.php#L73). If you want to give it a shot I think what you'd wan to try is to check the current element and if it's the one you're looking to push to a new page (based on ID, perhaps). I'm not sure yet how best to force a page break. Easiest may be to use the begin_frame event to update the element styling to force a page break if the current page is not even/odd (depending on your requirement).

**Comment by alex-webmaster** [comment]
This is my solution that works for me.
This allows new pages to be output from odd.
If you need output new pages from even, change '$PAGE_NUM%2 != 0' to '$PAGE_NUM%2 == 0'

```
$html ='<body>';
$html .='here is my html';

$html.='<script type="text/php">
                if ( isset($pdf) ) {
                    if($PAGE_NUM%2 != 0)  //add empty page if odd one
                    {
                        $pdf->new_page();
                    }
                }
            </script>';    
$html.='<div style="page-break-before: always;"></div>'; //start new page

$html .='this is my html which will start at odd page';
$html .='</body>';



$options = new Options();
$options->setIsRemoteEnabled(true);
    
$dompdf = new Dompdf($options);
$dompdf->loadHtml($html);
$dompdf->set_option("isPhpEnabled", true);

$dompdf->setPaper('A4', 'portrait');
$dompdf->render();

$dompdf->stream();
```
